lib/Partitioner/Partitioner.cpp

```diff
@@ -0,0 +1,365 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Partitioner/Partitioner.h"
+#include "glow/Graph/Utils.h"
+
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace glow;
+using llvm::isa;
+
+/// Margin for code itself -- should be adjusted later.
+#define MARGIN 1024 * 128
+
+struct BFSLevel {
+  /// A list of <level, nodelist> with BFS order.
+  std::vector<std::pair<int, std::vector<Node *>>> levels;
+  /// A set of visited nodes.
+  std::unordered_set<const Node *> visited;
+};
+
+Partitioner::Partitioner(Module *parent, std::vector<DeviceInfo> &devices)
+    : deviceInfo_(devices) {
+  module_ = parent;
+  // Check if all the functions are in the same structure.
+  if (!isSameFunctionFamily()) {
+    assert("not in the same function family");
+  }
+  memSize = module_->getConstantsSize();
+}
+
+/// Get the represent function : the one with the largest input, and update the
+/// memSize.
+void Partitioner::selectRepFunc() {
+  auto funcList = module_->getFunctions();
+  for (Function *F : funcList) {
+    size_t size = memSize;
+
+    // The set to keep the placeholders (only for Inputs) whose size is already
+    // calcualted.
+    std::set<std::string> pSet;
+
+    for (auto &node : F->getNodes()) {
+      int n = node.getNumInputs();
+      if (node.getKind() == Kinded::Kind::SaveNodeKind) {
+        // Special node, the placehodler should be ignored?
+        continue;
+      }
+      for (int i = 0; i < n; i++) {
+        Placeholder *in =
+            llvm::dyn_cast<Placeholder>(node.getNthInput(i).getNode());
+        if (in && (pSet.find(in->getName()) == pSet.end())) {
+          auto ty = in->getType();
+          size += ty->getSizeInBytes();
+          pSet.insert(in->getName());
+        }
+      }
+    }
+    if (size > memSize) {
+      F_ = F;
+      memSize = size;
+    }
+  }
+}
+
+/// Get the minimal memory requirement (constant) for each op in the function.
+void Partitioner::initOpMemUsage() {
+  memUsage.clear();
+  for (auto &node : F_->getNodes()) {
+    int n = node.getNumInputs();
+    unsigned size = 0;
+    if (node.getKind() != Kinded::Kind::SaveNodeKind) {
```

Comment:
easier to read if we do `if (node.getKind() == Kinded::Kind::SaveNodeKind) {continue;}`

Reply:
Actually, there is a statement after this condition. I think it is OK to use "!=" here since I don't want to duplicate that statement .

Reply:
@beicy Use early exit here. 

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code